0010-endpoint-metadata-extension/README.md

```diff
@@ -0,0 +1,687 @@
+---
+name: Endpoint Metadata Extension
+shortname: 10/EPMDX
+status: draft
+editor: Volodymyr Tkhir <vtkhir@cybervisiontech.com>
+---
+
+- [Introduction](#introduction)
+- [Requirements and constraints](#requirements-and-constraints)
+  - [Metadata keys](#metadata-keys)
+  - [Metadata values](#metadata-values)
+  - [Supported operations](#supported-operations)
+- [Design](#design)
+  - [Full metadata update](#full-metadata-update)
+    - [Full metadata update request](#full-metadata-update-request)
+    - [Full metadata update response](#full-metadata-update-response)
+  - [Partial metadata update](#partial-metadata-update)
+    - [Partial metadata update request](#partial-metadata-update-request)
+    - [Partial metadata update response](#partial-metadata-update-response)
+  - [Get metadata](#get-metadata)
+    - [Get metadata request](#get-metadata-request)
+    - [Get metadata response](#get-metadata-response)
+  - [Get metadata keys](#get-metadata-keys)
+    - [Get metadata keys request](#get-metadata-keys-request)
+    - [Get metadata keys response](#get-metadata-keys-response)
+  - [Delete metadata key](#delete-metadata-key)
+    - [Delete metadata key request](#delete-metadata-key-request)
+    - [Delete metadata key response](#delete-metadata-key-response)
+
+
+## Introduction
+
+The Endpoint Metadata Extension (EPMDX) protocol is an endpoint-aware [Kaa Protocol](/0001-kaa-protocol/README.md) extension.
+
+It is intended to manage endpoint metadata.
+
+## Requirements and constraints
+
+### Metadata keys
+EPMDX protocol supports only string type of metadata keys.
+All UTF-8 characters are supported in metadata key names.
+Metadata key must contain at least one non-whitespace character.
```

Comment:
A little correction: `PUT api/v1/endpoints/{endpointId}/metadata/{metadataKey}` with empty `metadataKey` is `api/v1/endpoints/{endpointId}/metadata/`, not `api/v1/endpoints/{endpointId}/metadata//`. Anyway, both look fine for me.

As for UI, we don't have a requirement to make all arbitrary keys distinguishable. With Unicode, you сan't do that anyway. (Have you noticed that `с` in `сan't` is actually Cyrillic?)

A simple possibility does not mean the user will create _dozens_ of whitespace-only keys, so I don't care.

Ok, so here is an opposite proposal. Rather than make arbitrary limitations, let's take ones from [NATS](https://nats.io/documentation/internals/nats-protocol/#nats-protocol-conventions) (but probably forbid dots):
> Subject names: Subject names, including reply subject (INBOX) names, are case-sensitive and must be non-empty alphanumeric strings with no embedded whitespace, and optionally token-delimited using the dot character (.)

These limitations are perfect for data storages of any kind (and basically any system). As a bonus, this would allow us to publish metadata updates within NATS on a per-key basis.